KIALI-2245 gRPC support

jshaughn opened the pull request:
- no longer use 100% default for %traffic, it doesn't play well with Find
- change edge to a single traffic protocol, multiple protocols doesn't make sense.
- add grpc fields to Find and Find Help
- fix aggregation issues in graph summary
- fix kiali-2290 app box issue
- add 'protocol' field to cy edge data
- add 'requestProtocol' option to metrics queries
- KIALI-2296 Only show %traffic label for edges with traffic

THIS PR REQUIRES SERVER PR https://github.com/kiali/kiali/pull/807

jshaughn commented:
@israel-hdez Thanks for offering a review.  I know that at the moment there is no change to the legend because I didn't want to change the legend and then find out we were making changes to the edges.  If you want to see GRPC traffic without adding any services, you can look at istio-system.   Please take at all of the various side panels to make sure they are still in good shape for http and tcp, and also incorporate grpc as needed.

israel-hdez commented on src/pages/Graph/GraphHelpFind.tsx:
```diff
@@ -322,9 +324,12 @@ export default class GraphHelpFind extends React.Component<GraphHelpFindProps> {
+                        { id: 'ec20', c: '%grpctraffic <op> <number>', n: 'range: [0..100]' },
+                        { id: 'ec23', c: 'http <op> <number>', n: 'unit: requests per second' },
+                        { id: 'ec24', c: '%httperr <op> <number>', n: 'range: [0..100]' },
+                        { id: 'ec25', c: '%httptraffic <op> <number>', n: 'range: [0..100]' },
```
In the "Examples" a change from `%traffic` -> `%httptraffic` is missing

jshaughn replied:
Si, gracias.

israel-hdez commented on src/pages/Graph/SummaryPanelGraph.tsx:
```diff
@@ -77,15 +77,14 @@ export default class SummaryPanelGraph extends React.Component<SummaryPanelPropT
-    const baseQuery = cy
-      .nodes()
-      .filter('[!isGroup]')
-      .filter('[!isRoot]');
```
`!isRoot` filter is removed... I remember this was to not count "unknown". But yes, there are other cases where we want to count roots...

jshaughn replied:
Seeing your comment I looked at this code again and simplified it just a bit.  Each count is now just using a dedicated cy selector based on node type.  When dealing with app nodes I explicitly filter out the group nodes.  Please take a look, I think it is more clear this way.

israel-hdez commented on src/pages/Graph/SummaryPanelGraph.tsx:
```diff
@@ -77,15 +77,14 @@ export default class SummaryPanelGraph extends React.Component<SummaryPanelPropT
+    const numApps = baseQuery.filter(`[nodeType = "${NodeType.APP}"]`).size();
     const numEdges = cy.edges().size();
-    const trafficRate = getAccumulatedTrafficRate(cy.edges());
+    const nonServiceEdges = baseQuery.filter(`[nodeType != "${NodeType.SERVICE}"]`).edgesTo('*'); // don't count requests from injected service node
```
+1
good bugfix when counting traffic.

jshaughn replied:
Thanks,

israel-hdez commented on src/types/Graph.ts:
```diff
@@ -13,6 +13,13 @@ export interface SummaryData {
+  HTTP = 'http',
+  TCP = 'tcp'
+}
+export type ProtocolType = Protocol.GRPC | Protocol.HTTP | Protocol.TCP;
```
Is `ProtocolType` really necessary?
I think you could use `Protocol` enum to type variables.

jshaughn replied:
Oops.  +1, At first I had ProtocolType with the string literals and then I added the enum and basically forgot that I no longer needed the type.

abonas commented:
What kind of visual changes does this PR provide? Can you please add a screenshot?
IIRC grpc on the graph is still in UX discussions in
https://github.com/kiali/kiali-design/issues/84

jshaughn commented on src/components/SummaryPanel/InOutRateChart.tsx:
```diff
@@ -2,7 +2,86 @@ import * as React from 'react';
+          }
+        }}
+        data={{
+          groups: [['OK', '3xx', '4xx', '5xx']],
```
Yes, good catch.

jshaughn commented on src/pages/Graph/SummaryPanelNode.tsx:
```diff
@@ -280,20 +317,41 @@ export default class SummaryPanelNode extends React.Component<SummaryPanelPropTy
+          {this.hasHttpTraffic(node) && this.renderHttpRates(node)}
+          <div>{this.renderCharts(node)}</div>
+          {!this.hasGrpcTraffic(node) && renderNoTraffic('GRPC')}
+          {!this.hasHttpTraffic(node) && renderNoTraffic('HTTP')}
```
I agree, this doesn't look great for unknown. But the basic approach is to present as much telemetry data as possible and then follow it with messages about the protocols with no telemetry data.  I think this is still the best layout.  In this special case for unknown I added a horizontal rule to separate the two messages.  It looks a bit better to me.

jshaughn commented on src/store/Selectors/GraphData.ts:
```diff
@@ -69,9 +74,9 @@ const decorateGraphData = (graphData: any) => {
-            decoratedEdge.data = { ...protocol.rates, ...decoratedEdge.data };
-          });
+          const traffic = edge.data.traffic;
+          edge.data.traffic = undefined;
```
I added comments to the code to better explain what is happening here.

jshaughn commented:
> What kind of visual changes does this PR provide?

@abonas You can see the UX issue for screenshots: [kiali/kiali-design#84](https://github.com/kiali/kiali-design/issues/84)

jshaughn commented:
@israel-hdez I have pushed changes for your review.  Please take a look, Thanks!

israel-hdez commented on src/components/GraphFilter/__tests__/GraphFind.test.tsx:
```diff
@@ -67,21 +67,33 @@ describe('Parse find value test', () => {
     expect(instance.parseFindValue('outside')).toEqual('node[isOutside]');
     // @ts-ignore
-    expect(instance.parseFindValue('unused')).toEqual('node[isUnused]');
+    expect(instance.parseFindValue('outside')).toEqual('node[isOutside]');
```
Minor - I think you can remove this line. I cannot find any difference against the previous line.

jshaughn replied:
Good find, actually the second line is supposed to test 'outsider', it was missing the 'r',

israel-hdez commented on src/components/CytoscapeGraph/graphs/GraphStyles.ts:
```diff
@@ -103,7 +103,13 @@ export class GraphStyles {
+            if (pErr > 0) {
+              let sErr = pErr.toFixed(1);
+              sErr = `, ${sErr.endsWith('.0') ? pErr.toFixed(0) : sErr}%`;
+              content = rate.toFixed(2) + sErr;
```
Minor -- I think code will look simpler doing scaling:
```js
let sErr = Math.round(pErr * 10) / 10;
content = `${rate.toFixed(2)}, ${sErr.toString()}%`;
```
BUT... I'm not sure if this could fall in [floating point precision problems](https://stackoverflow.com/questions/588004/is-floating-point-math-broken) and display lots of digits. Probably, your version is safer for the UI, but scaling is easier to read for devs.

So, it's up to you.

jshaughn replied:
It's working, I'm OK as is.

israel-hdez commented on src/components/CytoscapeGraph/graphs/GraphStyles.ts:
```diff
@@ -115,20 +121,26 @@ export class GraphStyles {
-          content = pReq < 100.0 ? pReq.toFixed(1) + '%' : '100%';
+          if (pReq > 0) {
+            const sReq = pReq.toFixed(1);
+            content = `${sReq.endsWith('.0') ? pReq.toFixed(0) : sReq}%`;
```
Similar comment here (for scaling)

jshaughn replied:
I'm still OK :)

israel-hdez commented on src/pages/Graph/SummaryPanelNode.tsx:
```diff
@@ -360,10 +453,17 @@ export default class SummaryPanelNode extends React.Component<SummaryPanelPropTy
+                <Icon type="pf" name="info" /> Traffic to istio-system not included. Use edge for details.
+              </div>
+            </>
+          )}
```
Yikes! I see that the edge also doesn't have sparklines. I guess the edge also collects data from source telemetry.
But this message is OK. The edge, at least, will show in the rates-table the specific details.
This is just a comment :)

jshaughn replied:
We can chase istio-system corner cases for a long time, instead let's stop and wait for any negative feedback. Good testing.

jshaughn commented:
@mattmahoneyrh I have pushed changes to the PR.  I apologize that this is sort of becoming an uber-PR.  The Graph Find issue you pointed out above is not really related to this PR but was a known issue (at least to me).  As you said, it was technically correct but not particularly useful.  I made a relatively big change and I think it's now fixed.  To solve the issue I changed default values (from '0' to 'NaN') for our numeric fields in the graph elements.  Please re-check the side panels to make sure they look good and we still see the expected charts and sparklines, and we don't see any NaN values creeping in.

Additionally, please re-check the edge sparkline in the "istio-system destination node" case we talked about.  I think I fixed that issue as well, and now you should see a sparkline chart (as opposed to the "Not enough data..." message).

Finally, I made a couple of additions to the Find/Hide DSL.  You can now identify edges via protocol.  For example, "protocol = grpc".   Also, you can test for no activity by testing for NaN.  For example: "httpout = NaN".  (or any activity with != NaN).

mattmahoneyrh commented:
The "Not enough traffic..." msg is still seen:

![image](https://user-images.githubusercontent.com/7262987/52073490-a9a32500-2555-11e9-8a34-b8504c3a631d.png)

jshaughn commented:
> The "Not enough traffic..." msg is still seen:

OK, after meeting with @mattmahoneyrh we figured out what was happening here.  The edge sparkline issue was resolved but traffic to istio services doesn't happen on the same schedule as regular app traffic.  There can be lulls such that for the "Last Minute" there really isn't any traffic between the envoy proxy and the istio service, even if there is plenty of app traffic being processed.  By expanding to "Last 5 minutes" or greater we saw the sparklines, and they showed the dips in traffic that resulted in the lack of a sparkline for the short 1 minute duration.
